oak-run/src/main/java/org/apache/jackrabbit/oak/plugins/document/check/OrphanedNodeCheck.java

```diff
@@ -0,0 +1,159 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.jackrabbit.oak.plugins.document.check;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.jackrabbit.oak.commons.concurrent.ExecutorCloser;
+import org.apache.jackrabbit.oak.plugins.document.DocumentNodeState;
+import org.apache.jackrabbit.oak.plugins.document.DocumentNodeStore;
+import org.apache.jackrabbit.oak.plugins.document.NodeDocument;
+import org.apache.jackrabbit.oak.plugins.document.Path;
+import org.apache.jackrabbit.oak.plugins.document.RevisionVector;
+import org.apache.jackrabbit.oak.spi.state.NodeState;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * <code>OrphanedNodeCheck</code>...
+ */
+public class OrphanedNodeCheck implements DocumentProcessor, Closeable {
+
+    private final DocumentNodeStore ns;
+
+    private final RevisionVector headRevision;
+
+    private final ExecutorService executorService;
+
+    public OrphanedNodeCheck(DocumentNodeStore ns,
+                             RevisionVector headRevision,
+                             int numThread) {
+        this.ns = ns;
+        this.headRevision = headRevision;
+        this.executorService = new ThreadPoolExecutor(
+                numThread, numThread, 1, TimeUnit.MINUTES,
+                new LinkedBlockingQueue<>(1000),
+                new ThreadPoolExecutor.CallerRunsPolicy()
+        );
+    }
+
+    @Override
+    public void processDocument(@NotNull NodeDocument document,
+                                @NotNull BlockingQueue<Result> results) {
+        if (!document.isSplitDocument()) {
+            executorService.submit(new CheckDocument(ns, headRevision, document, results));
+        }
+    }
+
+    @Override
+    public void end(@NotNull BlockingQueue<Result> results)
+            throws InterruptedException {
+        executorService.shutdown();
+        if (!executorService.awaitTermination(5, TimeUnit.MINUTES)) {
+            String msg = "Checks still not finished after the last one has been submitted 5 minutes ago";
+            results.put(() -> {
+                StringBuilder sb = new StringBuilder("{");
```

Comment:
Can we please use JsonObject to create JSON?

Reply:
For better or worse, Oak already has it's own JSON library. IMHO that should be used.